File: dsa/week(3)/Operateroverload.py
from math import gcd

class Fraction:
    def __init__(self, numerator, denominator):
        self.numerator = numerator
        self.denominator = denominator

    def print(self):
        print(f"{self.numerator} / {self.denominator}")

    def simplify(self):
        common = gcd(self.numerator, self.denominator)
        self.numerator //= common
        self.denominator //= common

    def add(self, f2):
        lcm = self.denominator * f2.denominator
        x = lcm // self.denominator
        y = lcm // f2.denominator

        num = x * self.numerator + y * f2.numerator
        result = Fraction(num, lcm)
        result.simplify()
        return result

    def multiply(self, f2):
        num = self.numerator * f2.numerator
        denom = self.denominator * f2.denominator
        result = Fraction(num, denom)
        result.simplify()
        return result

    def equals(self, f2):
        self_simplified = Fraction(self.numerator, self.denominator)
        self_simplified.simplify()

        f2_simplified = Fraction(f2.numerator, f2.denominator)
        f2_simplified.simplify()

        return self_simplified.numerator == f2_simplified.numerator and self_simplified.denominator == f2_simplified.denominator

# Example usage

f1 = Fraction(10, 2)
f2 = Fraction(15, 4)

f3 = f1.add(f2)
f4 = f1.multiply(f2)

f1.print()
f2.print()
f3.print()
f4.print()

if f1.equals(f2):
    print("Equal")
else:
    print("Not equal")

File: dsa/week(3)/Operateroverload.js
class Fraction {
    constructor(numerator, denominator) {
        this.numerator = numerator;
        this.denominator = denominator;
    }

    print() {
        console.log(`${this.numerator} / ${this.denominator}`);
    }

    simplify() {
        const gcd = this.#findGCD(this.numerator, this.denominator);
        this.numerator = this.numerator / gcd;
        this.denominator = this.denominator / gcd;
    }

    #findGCD(a, b) {
        return b === 0 ? a : this.#findGCD(b, a % b);
    }

    add(f2) {
        const lcm = this.denominator * f2.denominator;
        const x = lcm / this.denominator;
        const y = lcm / f2.denominator;

        const num = x * this.numerator + y * f2.numerator;
        const result = new Fraction(num, lcm);
        result.simplify();
        return result;
    }

    multiply(f2) {
        const num = this.numerator * f2.numerator;
        const denom = this.denominator * f2.denominator;
        const result = new Fraction(num, denom);
        result.simplify();
        return result;
    }

    equals(f2) {
        const thisSimplified = new Fraction(this.numerator, this.denominator);
        thisSimplified.simplify();

        const f2Simplified = new Fraction(f2.numerator, f2.denominator);
        f2Simplified.simplify();

        return thisSimplified.numerator === f2Simplified.numerator &&
               thisSimplified.denominator === f2Simplified.denominator;
    }
}

// Example usage
const f1 = new Fraction(10, 2);
const f2 = new Fraction(15, 4);

const f3 = f1.add(f2);
const f4 = f1.multiply(f2);

f1.print();
f2.print();
f3.print();
f4.print();

console.log(f1.equals(f2) ? "Equal" : "Not equal");
